src/logging.h

```diff
@@ -94,6 +97,13 @@ namespace BCLog {
         /** Log categories bitfield. */
         std::atomic<uint32_t> m_categories{0};
 
+        //! Category-specific log level. Overrides the global `m_log_level`.
+        std::unordered_map<LogFlags, Level> m_category_log_levels;
+
+        //! If there is no category-specific log level, all logs with a severity
+        //! level lower than `m_log_level` will be ignored.
+        Level m_log_level{Level::Info};
```

Comment:
If the newly added `m_category_log_levels` and `m_log_level` can be accessed concurrently from multiple threads (I think they can), then they should be either `atomic`, `const` (and set in the constructor) or protected by a mutex. Similarly to the neighboring e.g. `m_categories`.

Reply:
Data members shouldn't be const.  These are set once on startup and then not mutated. ISTM if changing them during runtime is enabled then they need to be thread-safe.

Reply:
Looking into this, `m_log_level` can be `atomic`, but since `m_category_log_levels` isn't a trivially copyable type, I wonder if we can guard it with the existing `Logger::m_cs`.

Reply:
Yes, `m_cs` should work.

I looked into making these `const` and setting them from the constructor, but that seems nearly impossible with the current `LogInstance()` and `g_logger` (we don't have the args at that time). Significant changes would be required to achieve that.

Reply:
Done (I think).

Reply:
Suggestions feel free to ignore:

I think a `std::shared_mutex` will be a good candidate for this situation. It will allow us not to need to synchronize calls to `LogPrint` (Reader) and only synchronize calls when we occasionally modify (Writer) `m_category_log_levels` during initialization or runtime during modification (future plan).

However, this might very well be a premature optimization as `m_cs` is only held for a very short period in `SetCategoryLogLevel` and `CategoryLevels`.